### PR TITLE
[autopilot] distribute available funds among channels

### PR DESCRIPTION
**Pull request by halseth**
This commit fixes a regression in how we allocate funds to attempted
channels. We would earlier stay within the channel size limits, but we
wouldn't account for funds consumed by other channels being opened in
parallel.

We fix this by introducing a loop which greadily tries to distribute the
funds among the channels to open, and reduces the number of channels to
open in case not enough funds are available to satisfy the channel size
limits.

Fixes https://github.com/lightningnetwork/lnd/issues/2609

### CODE REVIEW

File: autopilot/agent.go [hunk 1]
```diff
@@ -565,10 +565,29 @@ func (a *Agent) openChans(availableFunds btcutil.Amount, numChans uint32,
...
 
-	// As channel size we'll use the maximum channel size available.
-	chanSize := a.cfg.Constraints.MaxChanSize()
-	if availableFunds-chanSize < 0 {
-		chanSize = availableFunds
+	// We'll keep decreasing the number of channel we will open until we
+	// have enough funds available for all of them.
+	var chanSize btcutil.Amount
+	for ; numChans > 0; numChans-- {
+		// We'll start by attempting to distribute the funds evenly
```
Comment (Roasbeef) [inline autopilot/agent.go]: The existing version of this loop would instead (more greedily) attempt to _maximize_ the size of each channel respecting the min and max channel limits. Pre-amp, I think this might be a better approach, as it'll mean they'll be able to send/receive more of their deposited funds in a single go. With the new atpl selection taking into account node availability data on mainnet, it'll also be less likely that they have a single larger channel to a lack luster node.

Comment (Roasbeef) [reply autopilot/agent.go]: I still also think we should _attach the recommended channel size_ to node score. The node score can then accept the constraints and use that to make decision w.r.t channel size. Externally, we can manage things like the total number of channels to open. It's still not clear to me _what_ exactly we'd add to the existing channel constraints to allow it to recommend channel sizes in a more intelligent manner. It seems much more natural that node selection be contingent on how large of channels we want to open, and how much we have available at our disposal.

Comment (halseth) [reply autopilot/agent.go]: Agree on trying to open as large channels as possible 👍 

When getting scores, it already takes the wanted channel size into account (that is, it is part of the API, the current heuristics ignore it). Isn't that what we want? 

Having the heuristic return return a score+size for each node makes it hard to reason about where to open channels, as the agent has two sets of variables to consider.

Comment (Roasbeef) [reply autopilot/agent.go]: > When getting scores, it already takes the wanted channel size into account (that is, it is part of the API, the current heuristics ignore it). Isn't that what we want?

Yeah, that's the current behavior, but then it leaves open the question of how do we determine what channels size we want in the first place? If the scoring is aware of the total budget, then based on that it may actually score nodes differently. A routing node looking to allocate 1 billion SAT within channels may warrant a distinct scoring from a client looking to open up 100 mil SAT worth of channels. Or as an example, if we're examining a node for scoring, and the max flow from/to the node is 1 mil SAT, then we may not want to open a 10 billion sat channel to it, as we'd only be able to ever effectively use a very small % of it. 

I brought this up when the new channel constraints struct was introduced, as it seemed to attempt to side step this fundamental problem rather than take it on head first. In a world of splicing, maybe this becomes less important, as we only care about our min channel size to establish connectivity, then we can slowly pump up the channels over time in a reactive manner. 

> Having the heuristic return return a score+size for each node makes it hard to reason about where to open channels, as the agent has two sets of variables to consider.

True it makes things a bit more involved. The question then is, is it better to actually confront this problem, or delegate things to a more advanced channel constraint system that has yet to be developed or researched? I don't have a clear idea of how that system would even work. Would it consider the level of utilization of the current channels, and attempt to select an ideal channel size from that? Would it aim only open channels that are powers of two? Would it first do a dry run node scoring and then tune the channel size from there on? 

Not saying we need to tackle this issue at this very moment. Just want to ensure we're not forgetting about the other important component aside from who to open channels to: how large should those channels be?

Comment (halseth) [reply autopilot/agent.go]: We can model the "perfect channel size" as a separate heuristic. That way the agent would go through steps 
`1) with my current setup determine a size for my next channel 2) given this size, determine which node to open a channel to.`

An alternative flow could be 
`1) Given my current channels, determine the best node on the network to open a channel to 2) given this node, what is the best channel size`

Modeling these decisions as a series of individual heuristics makes it easier to combine and improve on the various parts. In the current case it would simply be 1) max channel size 2) use bos score list to find best node.

Moving all this logic into the heuristic would totally be possible fwiw, then the API would look more like the previous `Select` API: give me a (set of) channel(s) to open. Behind this the scoring and channel size logic would need to be performed.

If we get these steps right, I imagine we can find the "optimal" channel by (binary/ternary) searching for a size, getting scores for this size, then choosing the size that gives the highest score. I think this will be further down the road though, as at this point I think choosing the channel size independently of the score is plenty good enough.

Comment (Roasbeef) [reply autopilot/agent.go]: Alright, with that all said and done, actionable step here seems to be: make the loop greedily try to maximize channel size as it did before the switch to scoring based heuristics.

Comment (halseth) [reply autopilot/agent.go]: Fixed: https://github.com/lightningnetwork/lnd/pull/2633/commits/ac1102ca109c19551a04c01606ccc4e60c3d9945

**Comment by halseth** [comment]
PTAL @Roasbeef

File: autopilot/agent.go [hunk 2]
```diff
@@ -799,6 +817,9 @@ func (a *Agent) executeDirective(directive AttachmentDirective) {
 
 	// Since the channel open was successful and is currently pending,
 	// we'll trigger the autopilot agent to query for more peers.
+	// TODO(halseth): this triggers a new loop before all the new channels
```
[the remark this answers is not in the document]
Comment (halseth) [reply autopilot/agent.go]: Yes, but this happens in a goroutine, and they all get spawned in parallel, so we have no guarantee that they are all added to the map before the first one finishes.

File: autopilot/agent_test.go [hunk 3]
```diff
@@ -1699,3 +1699,390 @@ func TestAgentQuitWhenPendingConns(t *testing.T) {
...
+	select {
+	case constraints.moreChansResps <- moreChansResp{
+		numMore: 3,
+		amt:     channelBudget,
+	}:
+	case <-time.After(time.Second * 10):
+		t.Fatalf("heuristic wasn't queried in time")
+	}
+
+	// Since the maximum channel size is 1 BTC, the agent won't have enough
```
Comment (Roasbeef) [inline autopilot/agent_test.go]: I find these tests hard to follow at times in general. I think we can reduce some vertical space here with usage of some a helper functions to handle sending/receiving messages to/from the various mocked interfaces. I'd expect the logic to go "ok I have 1.5 BTC and the max size is 1 BTC, so I'll open to channels here for 1 and 0.5 BTC". Instead, as is, it'll instead execute the 1 BTC and the 0.5 BTC during distinct iterations. 

This is a regression compared to the behavior of the prior system, that would attempt to pipeline open as many channels as possible.

Comment (halseth) [reply autopilot/agent_test.go]: The rationale behind doing it this way was that you don't know if an opening attempt will always succeed, so I would wait until the next iteration to have more information about which ones succeed and potentially open a bigger channel.

Looking at it now though, I see that it won't matter much, since we will eventually end up with the smaller channel anyway, so whether it is being opened last or not makes no difference. Will fix!

Also agree on the verbosity of the tests, will clean they up a bit.

**Comment by Roasbeef** [comment]
Pong.

**Comment by halseth** [comment]
Comments addressed. PTAL @Roasbeef

File: autopilot/agent.go [hunk 4]
```diff
@@ -567,10 +567,15 @@ func (a *Agent) openChans(availableFunds btcutil.Amount, numChans uint32,
 
 	// As channel size we'll use the maximum channel size available.
 	chanSize := a.cfg.Constraints.MaxChanSize()
-	if availableFunds-chanSize < 0 {
+	if availableFunds < chanSize {
 		chanSize = availableFunds
 	}
 
+	if chanSize < a.cfg.Constraints.MinChanSize() {
```
Comment (wpaulino) [inline autopilot/agent.go]: We could move this up to avoid iterating through the whole graph in the event that we don't have enough funds.

File: autopilot/agent_test.go [hunk 5]
```diff
@@ -195,57 +222,95 @@ func TestAgentChannelOpenSignal(t *testing.T) {
...
+	}
 
-	// We'll send an initial "no" response to advance the agent past its
-	// initial check.
+	return ctx, cleanup
+}
+
+// respondMoreChans consumes the moreChanArgs element and responds to the agent
+// with the given moreChansResp.
+func respondMoreChans(t *testing.T, testCtx *testContext, resp moreChansResp) {
```
Comment (wpaulino) [inline autopilot/agent_test.go]: Nit: can use `t.Helper` here as well as in the rest of the assertion functions.

File: autopilot/agent_test.go [hunk 6]
```diff
@@ -372,77 +377,24 @@ func TestAgentChannelFailureSignal(t *testing.T) {
...
 	}
 
-	select {
-	case heuristic.nodeScoresResps <- map[NodeID]*NodeScore{
-		NewNodeID(node): fakeDirective,
-	}:
-	case <-time.After(time.Second * 10):
-		t.Fatal("heuristic wasn't queried in time")
-	}
+	respondNodeScores(t, testCtx,
```
Comment (wpaulino) [inline autopilot/agent_test.go]: Nit: should be of the form
```
respondNodeScores(
    ..., ...,
)
```

File: autopilot/agent_test.go [hunk 7]
```diff
@@ -150,10 +150,20 @@ func (m *mockChanController) SpliceOut(chanPoint *wire.OutPoint,
 
 var _ ChannelController = (*mockChanController)(nil)
 
-// TestAgentChannelOpenSignal tests that upon receipt of a chanOpenUpdate, then
-// agent modifies its local state accordingly, and reconsults the heuristic.
-func TestAgentChannelOpenSignal(t *testing.T) {
-	t.Parallel()
+type testContext struct {
```
Comment (wpaulino) [inline autopilot/agent_test.go]: 💯

File: autopilot/agent_test.go [hunk 8]
```diff
@@ -1091,3 +1091,219 @@ func TestAgentQuitWhenPendingConns(t *testing.T) {
...
+		// All nodes in the graph should be potential channel
+		// candidates.
+		if len(req.nodes) != len(nodeScores) {
+			t.Fatalf("expected %v nodes, instead had %v",
+				len(nodeScores), len(req.nodes))
+		}
+
+		// One channel is already open.
+		if len(req.chans) != int(existingChans) {
+			t.Fatalf("expected 1 existing channel, got %v",
```
Comment (wpaulino) [inline autopilot/agent_test.go]: Should be expected `existingChans`?

File: autopilot/agent_test.go [hunk 9]
```diff
@@ -1091,3 +1091,219 @@ func TestAgentQuitWhenPendingConns(t *testing.T) {
...
+			Score:  0.5,
+		}
+	}
+
+	// The agent should now query the heuristic in order to determine its
+	// next action as it local state has now been modified.
+	select {
+	case arg := <-testCtx.constraints.moreChanArgs:
+		if len(arg.chans) != 0 {
+			t.Fatalf("expected agent to have 1 channel already "+
```
Comment (wpaulino) [inline autopilot/agent_test.go]: Should be `expected agent to not have any open channels`?

File: autopilot/agent.go (hunk 4 above)
Comment (halseth) [reply autopilot/agent.go]: It is already checked before calling `openChans`, this is just an additional sanity check :P

File: autopilot/agent_test.go (hunk 5 above)
Comment (halseth) [reply autopilot/agent_test.go]: Fixed: https://github.com/lightningnetwork/lnd/pull/2633/commits/8d148e78a7fcf865789724f573d97aa91ea1b537

File: autopilot/agent_test.go (hunk 6 above)
Comment (halseth) [reply autopilot/agent_test.go]: Fixed: https://github.com/lightningnetwork/lnd/pull/2633/commits/28f79b54069838fdabf89f069214cd203ed6c4bc

File: autopilot/agent_test.go (hunk 8 above)
Comment (halseth) [reply autopilot/agent_test.go]: Fixed: https://github.com/lightningnetwork/lnd/pull/2633/commits/c201211ed2772ecb6946e224177eed0812697405

File: autopilot/agent_test.go (hunk 9 above)
Comment (halseth) [reply autopilot/agent_test.go]: Fixed: https://github.com/lightningnetwork/lnd/pull/2633/commits/30bf23d9256a5d2dec7fb980b23bf9434c3196e0

**Comment by halseth** [comment]
Comments addressed and rebase PTAL @wpaulino @Roasbeef

File: autopilot/agent_test.go [hunk 10]
```diff
@@ -195,57 +222,95 @@ func TestAgentChannelOpenSignal(t *testing.T) {
...
+	select {
+	case testCtx.constraints.moreChansResps <- resp:
+	case <-time.After(time.Second * 10):
+		t.Fatalf("response wasn't sent in time")
+	}
+}
+
+// respondMoreChans consumes the nodeScoresArgs element and responds to the
+// agent with the given node scores.
+func respondNodeScores(t *testing.T, testCtx *testContext,
```
Comment (Roasbeef) [inline autopilot/agent_test.go]: 👍

**Comment by halseth** [comment]
Rebased.
